src/algebra/symmetrized.lean

```diff
@@ -0,0 +1,216 @@
+not_congr $ unsym_eq_one_iff a
+
+@[to_additive] lemma sym_ne_one_iff [has_one α] (a : α) : sym a ≠ (1 : αˢʸᵐ) ↔ a ≠ (1 : α) :=
+not_congr $ sym_eq_one_iff a
+
+instance [add_comm_semigroup α] : add_comm_semigroup (αˢʸᵐ) :=
+unsym_injective.add_comm_semigroup _ (λ _ _, rfl)
+
+instance [add_monoid α] : add_monoid (αˢʸᵐ) :=
+unsym_injective.add_monoid_smul _ rfl (λ _ _, rfl) (λ _ _, rfl)
+
+instance [add_group α] : add_group (αˢʸᵐ) :=
+unsym_injective.add_group_smul _ rfl
+  (λ _ _, rfl) (λ _, rfl) (λ _ _, rfl) (λ _ _, rfl) (λ _ _, rfl)
+
+instance [add_comm_monoid α] : add_comm_monoid (αˢʸᵐ) :=
+{ ..sym_alg.add_comm_semigroup, ..sym_alg.add_monoid }
+
+instance [add_comm_group α] : add_comm_group (αˢʸᵐ) :=
+{ ..sym_alg.add_comm_monoid, ..sym_alg.add_group }
+
+instance {R : Type*} [semiring R] [add_comm_monoid α] [module R α] : module R αˢʸᵐ :=
+function.injective.module R ⟨unsym, rfl, λ _ _, rfl⟩ (λ _ _, id) (λ _ _, rfl)
+
+lemma mul_def [ring α] [invertible (2 : α)] (a b : αˢʸᵐ) :
+  a * b = sym (⅟2*(unsym a * unsym b + unsym b * unsym a)) := by refl
+
+instance [has_mul α] [has_add α] [has_one α] [invertible (2 : α)] (a : α) [invertible a] :
+  invertible (sym a) :=
+{ inv_of := sym ⅟a,
+  inv_of_mul_self := begin
+    change ⅟2 * (⅟a * a + a * ⅟a) = 1,
+    rw [mul_inv_of_self, inv_of_mul_self],
+    exact mul_inv_of_self (⅟ 2),
+  end,
+  mul_inv_of_self := begin
+    change ⅟2 * (a* ⅟a + ⅟a * a)=1,
+    rw [mul_inv_of_self, inv_of_mul_self],
+    exact mul_inv_of_self (⅟ 2),
+  end }
+
+lemma unsym_mul [has_mul α] [has_add α] [has_one α] [invertible (2 : α)] (a b : αˢʸᵐ) :
+  unsym (a * b) = ⅟2*(unsym a * unsym b + unsym b * unsym a) := by refl
+
+lemma sym_mul_sym [has_mul α] [has_add α] [has_one α] [invertible (2 : α)] (a b : α) :
+  sym a * sym b = sym (⅟2*(a * b + b * a)) :=
+rfl
+
+/- The symmetrization of a real (unital, associative) algebra is a non-associative ring -/
+instance [ring α] [invertible (2 : α)] : non_unital_non_assoc_ring (αˢʸᵐ) :=
+{ zero_mul := λ _,
+  begin
+    rw [mul_def, unsym_zero, zero_mul, mul_zero, add_zero, mul_zero, sym_zero],
+    exact rfl,
+  end,
+  mul_zero :=  λ _,
+  begin
+    rw [mul_def, unsym_zero, zero_mul, mul_zero, add_zero, mul_zero, sym_zero],
+    exact rfl,
+  end,
+  left_distrib := λ a b c, match a, b, c with
+    | sym a, sym b, sym c := begin
+      rw [sym_mul_sym, sym_mul_sym, ←sym_add, sym_mul_sym, ←sym_add, mul_add a, add_mul _ _ a,
+        add_add_add_comm, mul_add],
+    end
+  end,
+  right_distrib := λ a b c, match a, b, c with
+    | sym a, sym b, sym c := begin
+      rw [sym_mul_sym, sym_mul_sym, ←sym_add, sym_mul_sym, ←sym_add, mul_add c, add_mul _ _ c,
+        add_add_add_comm, mul_add],
+    end
+  end,
+  ..sym_alg.has_mul,
+  ..sym_alg.add_comm_group, }
+
+/- The squaring operation coincides for both multiplications -/
+lemma sym_squares [ring α] [invertible (2 : α)] (a : αˢʸᵐ) : unsym(a*a) = unsym a * unsym a :=
```

Comment:
```suggestion
lemma unsym_mul_self [ring α] [invertible (2 : α)] (a : αˢʸᵐ) : unsym (a*a) = unsym a * unsym a :=
```
or
```suggestion
lemma unsym_sq [ring α] [invertible (2 : α)] (a : αˢʸᵐ) : unsym (a^2) = unsym a ^ 2 :=
```
It's probably sensible to have both.

Reply:
I guess you should probably have `sym_sq` and `sym_mul_self` to match.

Reply:
I've added `sym_mul_self` for now.